### PR TITLE
New Adapter: Logan

### PR DESCRIPTION
**Pull request by WlsLogan**
doc - https://github.com/prebid/prebid.github.io/pull/4238

### CODE REVIEW

File: adapters/logan/logan.go
Comment (guscarreon) [inline adapters/logan/logan.go:114]: If the length of `seatBid.Bid` is `m` and the length of `request.Imp` is `n`, we have a time complexity of `O(mn)` with the current approach. At the cost of `O(n)`space we can improve the time complexity to `O(m+n)` if we map the `request.Imp[i].ID`s to the index values:
```
105       var response openrtb2.BidResponse
106       if err := json.Unmarshal(responseData.Body, &response); err != nil {
107           return nil, []error{err}
108       }
109  
110       bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
111       bidResponse.Currency = response.Cur
    +
    +     impsMappedByID := make(map[string]int, len(request.Imp))
    +     for i := range request.Imp {
    +         impsMappedByID[request.Imp[i].ID] = i
    +     }
    +
112       for _, seatBid := range response.SeatBid {
113           for i := range seatBid.Bid {
114 -             bidType, err := getMediaTypeForImp(seatBid.Bid[i].ImpID, request.Imp)
    +             bidType, err := getMediaTypeForImp(seatBid.Bid[i].ImpID, request.Imp, impsMappedByID)
115               if err != nil {
116                   return nil, []error{err}
117               }
118  
119               b := &adapters.TypedBid{
120                   Bid:     &seatBid.Bid[i],
121                   BidType: bidType,
122               }
123               bidResponse.Bids = append(bidResponse.Bids, b)
124           }
125       }
126       return bidResponse, nil
127   }
128  
129 - func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
    + func getMediaTypeForImp(impID string, imps []openrtb2.Imp, impMap map[string]int) (openrtb_ext.BidType, error) {
130 -     for _, imp := range imps {
131 -         if imp.ID == impID {
    +     if index, found := impMap[impID]; found {
132               if imp.Banner != nil {
133                   return openrtb_ext.BidTypeBanner, nil
134               }
135               if imp.Video != nil {
136                   return openrtb_ext.BidTypeVideo, nil
137               }
138               if imp.Native != nil {
139                   return openrtb_ext.BidTypeNative, nil
140               }
141 -         }
142       }
143  
144       return "", &errortypes.BadInput{
145           Message: fmt.Sprintf("Failed to find impression \"%s\"", impID),
146       }
147   }
adapters/logan/logan.go
```

**Review by guscarreon — COMMENTED** [review]
Looks pretty good. Left a couple of nitpicks.

File: static/bidder-info/logan.yaml
Comment (guscarreon) [inline static/bidder-info/logan.yaml:18]: Is the URL correct? I tried the redirect in the browser and I'm getting a 404:
image

Comment (VeronikaSolovei9) [reply static/bidder-info/logan.yaml:18]: Same for me. Here is my test url: https://ssp-cookie.logan.ai/pserver?gdpr=&gdpr_consent=&us_privacy=&redir=https%3A%2F%2Fprebid.adnxs.com%2Fpbs%2Fv1%2Fsetuid%3Fbidder%3Dlogan%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%5BUID%5D

**Comment by guscarreon** [comment]
Do Globalsun and Logan communicate with instances of the same service? In other words, are [http://endpoint.globalsun.io/pserver](http://endpoint.globalsun.io/pserver) and [http://USeast2.logan.ai/pserver](http://USeast2.logan.ai/pserver) the same server under the hood? 

If so, there would be advantages to make one of them an alias. We would aviod adding unneccessary lines to the codebase, and would be easier to maintain. Thoughts?

**Comment by WlsLogan** [comment]
> Do Globalsun and Logan communicate with instances of the same service? In other words, are http://endpoint.globalsun.io/pserver and http://USeast2.logan.ai/pserver the same server under the hood?
> 
> If so, there would be advantages to make one of them an alias. We would aviod adding unneccessary lines to the codebase, and would be easier to maintain. Thoughts?

No, these adapters are not related to each other in any way.

File: adapters/logan/logan.go
Comment (VeronikaSolovei9) [inline adapters/logan/logan.go:55]: Does this look better to you(instead of L43-L55):
```
var errors []error
....


                currImp := originalImpSlice[i]
		request.Imp = []openrtb2.Imp{currImp}

		var bidderExt reqBodyExt
		if err := json.Unmarshal(currImp.Ext, &bidderExt); err != nil {
			errors = append(errors, &errortypes.BadInput{
				Message: err.Error(),
			})
			continue // or return
		}

		bidderExt.LoganBidderExt.Type = "publisher" // constant

		finalImpExt, err := json.Marshal(bidderExt)
		if err != nil {
			return nil, []error{err}
		}
``` 
This is a common approach for other bidders. Unit tests pass with this change. 

Please extract "publisher" to a constant.

File: adapters/logan/logan.go
Comment (VeronikaSolovei9) [inline adapters/logan/logan.go:40]: Can be simplified to `originalImpSlice := request.Imp`

File: adapters/logan/logan.go
Comment (VeronikaSolovei9) [inline adapters/logan/logan.go:131]: `imps []openrtb2.Imp,` is unused parameter

**Review by Sonali-More-Xandr — APPROVED** [review]
LGTM! Go to go after one more approval.

**Review by VeronikaSolovei9 — APPROVED** [review]

**Pull request merged**
